File: core/python/models/entity.py
"""Pydantic models for entities (companies, people, funds)."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class EntityCreate(BaseModel):
    """Data required to insert a new entity."""

    name: str
    type: Literal["company", "person", "fund"]
    cik: str | None = None
    ticker: str | None = None
    exchange: str | None = None
    metadata: dict = {}


class Entity(EntityCreate):
    """An entity as stored in the database."""

    id: int
    created_at: datetime
    updated_at: datetime

File: frontend/src/schemas/entity.ts
import { z } from "zod";

export const entityCreateSchema = z.object({
  name: z.string().min(1),
  type: z.enum(["company", "person", "fund"]),
  cik: z.string().nullable().default(null),
  ticker: z.string().nullable().default(null),
  exchange: z.string().nullable().default(null),
  metadata: z.record(z.string(), z.unknown()).default({}),
});

export const entitySchema = entityCreateSchema.extend({
  id: z.number().int(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type EntityCreate = z.infer<typeof entityCreateSchema>;
export type Entity = z.infer<typeof entitySchema>;
